File: adapters/koreader/plugin/hash.lua
local sha = require("hash.sha2")

local hash = {}

-- Compute SHA-256 of a file and return as "sha256_<hex>"
function hash.sha256(file_path)
    local f, err = io.open(file_path, "rb")
    if not f then
        return nil, err
    end

    -- get the append function for streaming SHA-256
    local append = sha.sha256()

    -- feed file in chunks
    while true do
        local chunk = f:read(8192)
        if not chunk then break end
        append(chunk)
    end

    f:close()

    -- get final digest
    local digest = append()
    return "sha256_" .. digest
end

return hash

File: engine/pkg/utils/hash.go
// Copyright (C) 2026 Aron Davids
// SPDX-License-Identifier: GPL-3.0-or-later
package utils

import (
	"crypto/sha256"
	"encoding/hex"
	"io"
	"os"
)

// HashFile computes the SHA-256 hash of the file at the given path
// and returns it as a lowercase hexadecimal string.
func HashFile(path string) (string, error) {
	file, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer file.Close()

	hash := sha256.New()
	if _, err := io.Copy(hash, file); err != nil {
		return "", err
	}

	return hex.EncodeToString(hash.Sum(nil)), nil
}
